File: src/main/java/com/github/pnemonic/game/lottery/LotteryResultsReader.java
package com.github.pnemonic.game.lottery;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Collections;
import java.util.List;

public abstract class LotteryResultsReader {

	public LotteryResultsReader() {
		super();
	}

	public List<LotteryRecord> parse(File file) throws IOException {
		List<LotteryRecord> records;
		InputStream in = null;
		try {
			in = new FileInputStream(file);
			records = parse(in);
		} finally {
			if (in != null) {
				try {
					in.close();
				} catch (Exception e) {
					// consume
				}
			}
		}
		Collections.sort(records);
		return records;
	}

	public abstract List<LotteryRecord> parse(InputStream in) throws IOException;

}

File: src/main/java/com/github/pnemonic/game/lottery/LotteryResultsReader.kt
package com.github.pnemonic.game.lottery

import com.github.pnemonic.csv.CSVLine
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream

typealias RecordCallback = (LotteryRecord) -> Unit

abstract class LotteryResultsReader {
    @Throws(IOException::class)
    fun parse(file: File): List<LotteryRecord> {
        val records: List<LotteryRecord>
        FileInputStream(file).use {
            records = parse(it)
        }
        return records
    }

    @Throws(IOException::class)
    abstract fun parse(input: InputStream): List<LotteryRecord>

    @Throws(IOException::class)
    fun parse(file: File, visitor: RecordCallback) {
        FileInputStream(file).use {
            parse(it, visitor)
        }
    }

    @Throws(IOException::class)
    abstract fun parse(input: InputStream, visitor: RecordCallback)
}